<a href="https://colab.research.google.com/github/Dharma-Ranganathan/AllAboutPython/blob/main/ML_Projects_Codes/Heart_Disease_Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Heart Disease Classification

Overview :

1. Dependencies
2. Data collection
3. Data Pre-processing
4. Model Selection
5. Model Training
6. Model Evaluation

Dependencies

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

Data collection

In [2]:
heart_disease = pd.read_csv('/content/drive/MyDrive/Colab_python/heart_disease_data.csv')
heart_disease.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

Data Pre-processing

In [3]:
# checking if null values exists

heart_disease.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [4]:
# rechecking
heart_disease.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

* Null values doesn't exists in our dataset.

In [5]:
# basic statistical report
heart_disease.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [6]:
# checking shape of dataset
heart_disease.shape

(303, 14)

In [7]:
# checking if dataset is in balanced state
s1 = heart_disease['target'] == 1
s2 = heart_disease['target'] == 0
print(s1.sum(),s2.sum(),sep=' • ')

165 • 138


In [8]:
# one more approach in one line as same as previous
heart_disease.value_counts('target')

target
1    165
0    138
Name: count, dtype: int64


* dataset in balanced state only
* splitting features and label seperately

In [9]:
feature = heart_disease.drop('target',axis=1)
label = heart_disease['target']

* drops 'target' column and axis 1 says it's belongs to column
* Enabling target column to label variable itself

In [10]:
# checking feature
feature.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [11]:
# checking label
label.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

Model Selection

In [12]:
x_train,x_test,y_train,y_test = train_test_split(feature,label,test_size=0.2, stratify=label,random_state=42)

In [13]:
# checking how many test datas were taken
print(feature.shape,x_train.shape,x_test.shape)

(303, 13) (242, 13) (61, 13)


* 242 datas for training
* 61 datas for testing

Model Training

In [14]:
# model - binary classification
# Logistic regression

model = LogisticRegression()

In [15]:
# fitting training dataset to model

model.fit(x_train,y_train)

/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

Model Evaluation and it's Reports

In [16]:
# prediction of seen data - training data

x_train_pred = model.predict(x_train)

# accuracy

train_score = accuracy_score(x_train_pred,y_train)

print(f'accuracy of X trained : {train_score:.2f}')
print(f'percentage : {(train_score * 100):.2f}%')
print(f'out of {x_train.shape[0]} trained data, {(train_score * x_train.shape[0]):.0f} were accurate')

accuracy of X trained : 0.84
percentage : 84.30%
out of 242 trained data, 204 were accurate


In [17]:
# prediction of unseen data - test data

x_test_pred = model.predict(x_test)

# accuracy

test_score = accuracy_score(x_test_pred,y_test)

print(f'accuracy of X test : {test_score:.2f}')
print(f'percentage : {(test_score * 100):.2f}%')
print(f'out of {x_test.shape[0]} test data, {(test_score * x_test.shape[0]):.0f} were accurate')

accuracy of X test : 0.80
percentage : 80.33%
out of 61 test data, 49 were accurate


Classification reports

In [18]:
# reports of unseen data

test_reports = classification_report(x_test_pred,y_test)

print(test_reports)

              precision    recall  f1-score   support

           0       0.68      0.86      0.76        22
           1       0.91      0.77      0.83        39

    accuracy                           0.80        61
   macro avg       0.79      0.82      0.80        61
weighted avg       0.83      0.80      0.81        61



Building predictive system

In [23]:

# for diseased data

diseased = (38,1,2,138,175,0,1,173,0,0,2,4,2)

diseased = np.asarray(diseased)

diseased = diseased.reshape(1,-1)

print(diseased)

# for not diseased data

not_diseased = ( 67,1,0,160,286,0,0,108,1,1.5,1,3,2 )

not_diseased = np.asarray(not_diseased,dtype=int)

not_diseased = not_diseased.reshape(1,-1)

print(not_diseased)

[[ 38   1   2 138 175   0   1 173   0   0   2   4   2]]
[[ 67   1   0 160 286   0   0 108   1   1   1   3   2]]


In [28]:
# function for predictive system

def predict_single(data):

    if data[0] == 0:

        print('A person has no heart disease')

    else:

        print('A person has heart disease')

In [34]:
# model prediction with predictive system.

diseased_pred = model.predict(diseased)

not_diseased_pred = model.predict(not_diseased)

# calling function

print(diseased_pred)
predict_single(diseased_pred)
print()

print(not_diseased_pred)
predict_single(not_diseased_pred)
print()

[1]
A person has heart disease

[0]
A person has no heart disease



/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


So far, model performance and prediction is good as expected...

Thank you...